src/main/java/org/kuali/kfs/module/ar/businessobject/TimeBasedBillingPeriod.java

```diff
@@ -0,0 +1,132 @@
+package org.kuali.kfs.module.ar.businessobject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.joda.time.DateTimeZone;
+import org.joda.time.LocalDate;
+import org.kuali.kfs.coa.businessobject.AccountingPeriod;
+import org.kuali.kfs.coa.service.AccountingPeriodService;
+import org.kuali.kfs.module.ar.ArConstants;
+
+import java.sql.Date;
+
+public class TimeBasedBillingPeriod extends BillingPeriod {
+    
+    /*
+     * CU Customization
+     */
+    private static final Logger LOG = LogManager.getLogger(TimeBasedBillingPeriod.class);
+    
+    public TimeBasedBillingPeriod(ArConstants.BillingFrequencyValues billingFrequency, Date awardStartDate, Date currentDate, Date lastBilledDate, AccountingPeriodService accountingPeriodService) {
+        super(billingFrequency, awardStartDate, currentDate, lastBilledDate, accountingPeriodService);
+    }
+
+    @Override
+    protected Date determineEndDateByFrequency() {
+        final AccountingPeriod accountingPeriod = findPreviousAccountingPeriod(currentDate);
+        return accountingPeriod.getUniversityFiscalPeriodEndDate();
+    }
+
+    protected Integer calculatePreviousPeriodByFrequency(Integer currentAccountingPeriodCode, int periodsInBillingFrequency) {
+        Integer previousAccountingPeriodCode;
+        final int subAmt = (currentAccountingPeriodCode % periodsInBillingFrequency) == 0 ? periodsInBillingFrequency : currentAccountingPeriodCode % periodsInBillingFrequency;
+
+        previousAccountingPeriodCode = currentAccountingPeriodCode - subAmt;
+        return previousAccountingPeriodCode;
+    }
+
+    @Override
+    protected boolean canThisBeBilledByBillingFrequency() {
+        if (ArConstants.BillingFrequencyValues.ANNUALLY.equals(billingFrequency) && accountingPeriodService.getByDate(lastBilledDate).getUniversityFiscalYear() >= accountingPeriodService.getByDate(currentDate).getUniversityFiscalYear()) {
+            return false;
+        } else if (StringUtils.equals(findPreviousAccountingPeriod(currentDate).getUniversityFiscalPeriodCode(), findPreviousAccountingPeriod(lastBilledDate).getUniversityFiscalPeriodCode()) &&
+            accountingPeriodService.getByDate(lastBilledDate).getUniversityFiscalYear().equals(accountingPeriodService.getByDate(currentDate).getUniversityFiscalYear())) {
+            return false;
+        }
+
+        return true;
+    }
+    
+    /*
+     * CU Customization
+     */
+    @Override
+    protected Date determineStartDateByFrequency() {
+        if (lastBilledDate == null) {
+            LOG.error("determineStartDateByFrequency, unable to calculate start date by frequency when the last billed date iis NULL");
+            throw new IllegalStateException("Can not calculate the start date when the last billed date is NULL");
+        }
+        
+        LocalDate localDateLastBilledDate = new LocalDate(lastBilledDate.getTime(), DateTimeZone.UTC);
```

Comment:
I am not sure I understand why these 3 lines of code are needed, could this return calculateNextDay(lastBilledDate)?

Reply:
good catch, I didn't notice that function.  I shall switch